src/main/java/com/google/step/datamanager/DealManagerDatastore.java

```diff
@@ -99,54 +139,168 @@ public Deal updateDeal(Deal deal) {
     return transformEntitytoDeal(dealEntity);
   }
 
+  /**
+   * Returns a Deal object transformed from a deal entity.
+   *
+   * @param dealEntity Deal entity.
+   * @return a Deal object transformed from the entity.
+   */
+  private Deal transformEntitytoDeal(Entity dealEntity) {
+    long id = dealEntity.getKey().getId();
+    String description = (String) dealEntity.getProperty("description");
+    String photoBlobkey = (String) dealEntity.getProperty("photoBlobkey");
+    String start = (String) dealEntity.getProperty("start");
+    String end = (String) dealEntity.getProperty("end");
+    String source = (String) dealEntity.getProperty("source");
+    long posterId = (long) dealEntity.getProperty("posterId");
+    long restaurantId = (long) dealEntity.getProperty("restaurantId");
+    String timestamp = (String) dealEntity.getProperty("timestamp");
+    return new Deal(
+        id, description, photoBlobkey, start, end, source, posterId, restaurantId, timestamp);
+  }
+
+  /** Retrieves deals posted by _ followed by user */
+  private List<Deal> getDealsPublishedByFollowedRestaurantsOrUsers(
+      long userId, String fieldName, String filterAttribute) {
+    List<Deal> dealResults = new ArrayList<>();
+    List<Long> idsOfFollowedFieldName = getFollowedSomething(userId, fieldName);
+    for (Long id : idsOfFollowedFieldName) {
+      Filter propertyFilter = new FilterPredicate(filterAttribute, FilterOperator.EQUAL, id);
+      Query query = new Query("Deal").setFilter(propertyFilter);
+      PreparedQuery pq = datastore.prepare(query);
+      for (Entity entity : pq.asIterable()) {
+        dealResults.add(readDeal(entity.getKey().getId()));
+      }
+    }
+    return dealResults;
+  }
+
+  /**
+   * This is copied from the follow manager method which is private, would it be better to make the
+   * original method public?
+   */
+  private List<Long> getFollowedSomething(long followerId, String fieldName) {
+    Filter userFilter = new FilterPredicate(FOLLOWER_FIELD_NAME, FilterOperator.EQUAL, followerId);
+    Filter otherFilter = new FilterPredicate(fieldName, FilterOperator.NOT_EQUAL, null);
+    Filter filter = CompositeFilterOperator.and(userFilter, otherFilter);
+    Query query = new Query("Follow").setFilter(filter);
```

Comment:
don't call from the "Follow" table, use the FollowManager instead, or add methods if you need. Basically, don't have 2 classes managing the same table. Else, if say in the future we want to change the table format, then there will be too many things dependent on it.

Reply:
yea but the original method in follow manager is private so im not sure if i should make it public? since u said the people can tamper with the fieldnames

Reply:
what you're doing now also like tampering with field names
you can first call `getFollowedRestaurantIds(userId)` from `FollowManager` to get a `list` of restaurant IDs
Then lookup the Deal table to find restaurant IDs
That way you don't have to touch the Follow table at all.